backend/ping.py

```diff
@@ -0,0 +1,178 @@
+    for e in _split(filled_endpoints):
+        r = _scan(e)
+        results.update(r)
+    return results
+
+
+def _get_number_of_open_file_descriptors():
+    return len(os.listdir('/proc/{0}/fd'.format(str(os.getpid()))))
+
+
+def _get_file_descriptor_resource_limit():
+    soft, hard = resource.getrlimit(resource.RLIMIT_NOFILE)
+    return soft
+
+
+def _create_socket_2(address_family, socket_type, protocol, address_tuple):
+    try:
+        sock = socket.socket(address_family, socket_type, protocol)
+    except socket.error:
+        return None, False
+    sock_fd = sock.fileno()
+    fd_flags = fcntl.fcntl(sock_fd, fcntl.F_GETFL)
+    fd_flags |= os.O_NONBLOCK
+    fcntl.fcntl(sock_fd, fcntl.F_SETFL, fd_flags)
+    try:
+        sock.connect(address_tuple)
+    except socket.error as e:
+        if e.errno == errno.EINPROGRESS:
+            return sock, False
+        else:
+            sock.close()
+            return None, False
+    else:
+        sock.close()
+        return None, True
+
+
+def _create_socket(struct_getaddrinfo):
+    return _create_socket_2(struct_getaddrinfo[0],
+                              struct_getaddrinfo[1],
+                              struct_getaddrinfo[2],
+                              struct_getaddrinfo[4])
+
+
+def _wait_for_any_change(socket_fds):
+    select_results = select.select(socket_fds,
+                                   socket_fds,
+                                   socket_fds,
+                                   None)
+    return list(set(itertools.chain.from_iterable(select_results)))
+
+
+def _check_connection(sock):
+    try:
+        sock.getpeername()
+    except socket.error as e:
+        if e.errno != errno.ENOTCONN:
+            raise
+        result = False
+    else:
+        result = True
+    finally:
+        sock.close()
+    return result
+
+
+def _scan(endpoints):
+    results = {}
+    sockets = {}
+    try:
+        for host, port in endpoints:
+            getaddrinfo_results = socket.getaddrinfo(host,
+                                                     port,
+                                                     socket.AF_UNSPEC,
+                                                     socket.SOCK_STREAM)
+            for r in getaddrinfo_results:
+                sock, is_open = _create_socket(r)
+                if is_open:
+                    results[host, port] = True
+                    break
+                elif sock:
+                    # Yes, a nested tuple.
+                    sockets[sock.fileno()] = sock, (host, port)
+                    break
+            # If not broken...
+            else:
+                results[host, port] = False
+        while sockets:
+            fds = _wait_for_any_change(sockets.keys())
+            for fd in fds:
+                sock, host_and_port = sockets[fd]
+                results[host_and_port] = _check_connection(sock)
+                del sockets[fd]
```

Comment:
are you sure this works? you are deleting keys from a dictionary while iterating over it...